### PR TITLE
Fixes #3082: Update topic tree (removing towers of hanoi) and replace all image and content links properly

### LINKED ISSUE #3082: Perseus exercise triage
branch: develop
steps to reproduce: run exercise and to go to hint  

This exercise works but the hint takes you to a video url which targets the Khan Academy mothership.  
http://127.0.0.1:8008/learn/khan/humanities/renaissance-reformation/early-renaissance1/sculpture-architecture-florence/brunelleschi-and-ghiberti-s-sacrifice-of-isaac-quiz/  

This appears to affect all of the assessment items used in this exercise..

### PR DESCRIPTION
**Pull request by jamalex**
As per #3082, this:

- removes the towers-of-hanoi exercise, since it's an iframe onto the KA site
- downloads a fuller set of images (includes *.jpeg and *.JPG, etc)
- replaces all links to other KA resources to KA Lite's local copy (or removes links+text if KA Lite doesn't have them)

### CODE REVIEW

File: kalite/contentload/management/commands/generate_assessment_zips.py [hunk 1]
```diff
@@ -100,25 +106,100 @@ def all_image_urls(items):
...
+    global CONTENT_BY_READABLE_ID
+    if not CONTENT_BY_READABLE_ID:
+        CONTENT_BY_READABLE_ID = dict([(c["readable_id"], c) for c in get_content_cache().values()])
+    try:
+        return CONTENT_BY_READABLE_ID[readable_id]
+    except KeyError:
+        return CONTENT_BY_READABLE_ID.get(re.sub("\-+", "-", readable_id).lower(), None)
+
+
+def list_all_exercises_with_bad_links():
```
Comment (aronasorman) [inline kalite/contentload/management/commands/generate_assessment_zips.py]: This function is quite heavy. This also seems to be a "one-off" function not meant to be used anywhere. Can you put a leading underscore to indicate that it's an internal function? Also gives us an excuse not to write a test for it.

File: kalite/contentload/management/commands/generate_assessment_zips.py [hunk 2]
```diff
@@ -100,25 +106,100 @@ def all_image_urls(items):
...
+    newitems = copy.deepcopy(items)
+
+    # loop over every item, and replace links in them to point to local resources, if available
+    for item in newitems.itervalues():
+        item['item_data'] = localize_content_links(item['item_data'])
+
+    return newitems
+
+
+def localize_content_links(item_data):
```
Comment (aronasorman) [inline kalite/contentload/management/commands/generate_assessment_zips.py]: Tests for this function please.

File: kalite/contentload/management/commands/generate_assessment_zips.py [hunk 3]
```diff
@@ -100,25 +106,100 @@ def all_image_urls(items):
...
+    item_data = localize_image_urls(item_data)
+    item_data = localize_content_links(item_data)
+    return item_data
+
+
+def _old_image_url_to_content_url(matchobj):
     return "/content/khan/%s" % matchobj.group("filename")
+
+
+def localize_all_content_links(items):
```
Comment (aronasorman) [inline kalite/contentload/management/commands/generate_assessment_zips.py]: Please write some tests for this.

**Comment by aronasorman** [comment]
There's also a single test failure related to the changes to this test module.

File: kalite/contentload/management/commands/generate_assessment_zips.py [hunk 4]
```diff
@@ -56,7 +60,9 @@ def download_urls(zf, urls):
 
     urls = set(urls)
 
-    pool = ThreadPool(5)
+    # TODO(jamalex): this can be changed to 5 or so during download, but multiple threads writing to the zip
+    # at the same time leads to corruption, so you'll then need to re-run with it set to 1
+    pool = ThreadPool(1)
     download_to_zip_func = lambda url: download_url_to_zip(zf, url)
```
Comment (aronasorman) [inline kalite/contentload/management/commands/generate_assessment_zips.py]: Interesting, I guess we can split the downloading part to be concurrent and let the writing-to-zip part just happen on the main thread. Can you make such a refactoring a todo here?

File: kalite/contentload/management/commands/generate_assessment_zips.py [hunk 5]
```diff
@@ -100,25 +106,100 @@ def all_image_urls(items):
     for _, v in items.iteritems():
 
         item_data = v["item_data"]
-        imgurlregex = r"https?://[\w\.\-\/]+\/(?P<filename>[\w\.\-]+\.(png|gif|jpg))"
+        imgurlregex = r"https?://[\w\.\-\/]+\/(?P<filename>[\w\.\-]+\.(png|gif|jpg|jpeg))"
```
Comment (aronasorman) [inline kalite/contentload/management/commands/generate_assessment_zips.py]: Didn't we also see `JPG` extensions too?

File: kalite/contentload/management/commands/generate_assessment_zips.py [hunk 6]
```diff
@@ -100,25 +106,100 @@ def all_image_urls(items):
...
 
-        for match in re.finditer(imgurlregex, item_data):
+        for match in re.finditer(imgurlregex, item_data, flags=re.IGNORECASE):
```
Comment (aronasorman) [inline kalite/contentload/management/commands/generate_assessment_zips.py]: Oh, `IGNORECASE`. Carry on.

File: kalite/contentload/management/commands/generate_assessment_zips.py (hunk 4 above)
Comment (jamalex) [inline kalite/contentload/management/commands/generate_assessment_zips.py]: Or would a mutex around the write operations perhaps help?

Comment (aronasorman) [inline kalite/contentload/management/commands/generate_assessment_zips.py]: Yeah, sounds good too.

File: kalite/contentload/management/commands/generate_assessment_zips.py [hunk 7]
```diff
@@ -56,14 +62,14 @@ def download_urls(zf, urls):
 
     urls = set(urls)
 
-    pool = ThreadPool(5)
+    pool = ThreadPool(10)
```
Comment (jamalex) [inline kalite/contentload/management/commands/generate_assessment_zips.py]: Looks like a mutex (see below) fixes it beautifully. Let there be concurrency! :confetti_ball:

**Comment by MCGallaspy** [comment]
This corrupted zipfile issue you address here is pretty critical.

File: kalite/contentload/management/commands/generate_assessment_zips.py [hunk 8]
```diff
@@ -6,17 +6,29 @@
...
 import kalite.version as version
 
+from kalite.topic_tools import get_content_cache, get_exercise_cache
+
 logging = settings.LOG
 
 ZIP_FILE_PATH = os.path.join(settings.PROJECT_PATH, "assessment_item_resources.zip")
 ASSESSMENT_ITEMS_PATH = os.path.join(settings.PROJECT_PATH, "..", "data", "khan", "assessmentitems.json")
 
+IMAGE_URL_REGEX = re.compile('https?://[\w\.\-\/]+\/(?P<filename>[\w\.\-]+\.(png|gif|jpg|jpeg))', flags=re.IGNORECASE)
```
Comment (jamalex) [inline kalite/contentload/management/commands/generate_assessment_zips.py]: Added the `IGNORECASE` here, instead, so I could remove it from `re.sub` below and thereby get 2.6 compatibility.

File: kalite/contentload/management/commands/generate_assessment_zips.py (hunk 3 above)
Comment (jamalex) [inline kalite/contentload/management/commands/generate_assessment_zips.py]: Done! https://github.com/learningequality/ka-lite/pull/3342/files#diff-c6c51d6e18589552eb095d057df7adfcR56

**Comment by jamalex** [comment]
> This corrupted zipfile issue you address here is pretty critical.

Yeah -- I had gotten around it yesterday by just switching to a single thread, but nice to regain that parallellism.

**Comment by jamalex** [comment]
@aronasorman Ok, I think I've addressed everything and more. Just waiting for tests to pass.

**Comment by aronasorman** [comment]
I'm itching to get this merged, as it's blocking my i18n testing.

**Comment by jamalex** [comment]
Feel free to merge it, as the only failing test (which should now be fixed) was an unnecessary one.

**Comment by aronasorman** [comment]
Tests pass YEY!

**Pull request merged**